pkg/docker/image_id.go

```diff
@@ -0,0 +1,90 @@
+package docker
+
+import (
+	"fmt"
+	"strings"
+)
+
+type EmptyTag struct{}
+type NameTag string
+type DigestTag string
+
+type ImageTag interface {
+	Separator() string
+	String() string
+}
+
+func (EmptyTag) Separator() string { return "" }
+func (EmptyTag) String() string    { return "" }
+
+func (NameTag) Separator() string { return ":" }
+
+func (n NameTag) String() string { return string(n) }
+
+func (DigestTag) Separator() string { return "@" }
+
+func (n DigestTag) String() string { return string(n) }
+
+func ParseImageTag(s string) (string, ImageTag) {
+	// Check for a digest, and if not look for a tag
+	if strings.Contains(s, "@") {
+		tag_split := strings.Split(s, "@")
+		return tag_split[0], DigestTag(tag_split[1])
+	} else if strings.Contains(s, ":") {
+		tag_split := strings.Split(s, ":")
+		return tag_split[0], NameTag(tag_split[1])
+	}
+
+	return s, EmptyTag{}
+}
+
+type ImageID struct {
+	// e.g. ghcr.io
+	repository string
+
+	// e.g. ubuntu or organisation/ubuntu or organisation/user/ubuntu
+	name string
+
+	// e.g. latest or kinetic or @sha256:digest-string
+	tag ImageTag
+}
+
+func NewImageID(imageID string) ImageID {
```

Comment:
Not implying I think this is wrong or anything, but would it be possible to re-use the existing docker code for parsing this information? e.g. https://github.com/moby/moby/blob/master/vendor/github.com/docker/distribution/reference/reference.go#L65

Reply:
Oh yeah, that'd make life a bit easier. We already have it as an indirect dependency of docker/docker so won't introduce any more dependencies.  I'll take a look at doing that.

Reply:
Followed this advice, found a bug in my parsing so definitely a good move.  The interface is kind of ugly so I haven't let it leak outside the ImageID module as we really just want to be able to pick apart sections of the image id, and it seems unnecessary to require knowledge of how to cast the docker module type to various subtypes.